msprime/simulations.py

```diff
@@ -1528,6 +1528,88 @@ def population_size_trajectory(self, steps):
             N_t[j] = N
         return N_t
 
+    def mean_coalescence_time(
+            self, num_samples, min_pop_size=1, steps=None, rtol=0.005, max_iter=12):
+        """
+        Compute the mean time until coalescence between lineages of two samples drawn
+        from the sample configuration specified in `num_samples`. This is done using
+        :meth:`coalescence_rate_trajectory
+        <.DemographyDebugger.coalescence_rate_trajectory>`
+        to compute the probability that the lineages have not yet coalesced by time `t`,
+        and using these to approximate :math:`E[T] = \\int_t^\\infty P(T > t) dt`,
+        where :math:`T` is the coalescence time. See
+        :meth:`coalescence_rate_trajectory
+        <.DemographyDebugger.coalescence_rate_trajectory>`
+        for more details.
+
+        :param list num_samples: A list of the same length as the number
+            of populations, so that `num_samples[j]` is the number of sampled
+            chromosomes in subpopulation `j`.
+        :param int min_pop_size: See :meth:`coalescence_rate_trajectory
+            <.DemographyDebugger.coalescence_rate_trajectory>`.
+        :param list steps: The time discretization to start out with (by default,
+            picks something based on epoch times).
+        :param float rtol: The relative tolerance to determine mean coalescence time
+            to (used to decide when to stop subdividing the steps).
+        :param int max_iter: The maximum number of times to subdivide the steps.
+        :return: The mean coalescence time (a number).
+        :rtype: float
+        """
+
+        def mean_time(steps, P):
+            # Mean is int_0^infty P(T > t) dt, which we estimate by discrete integration
+            # assuming that f(t) = P(T > t) is piecewise exponential:
+            # if f(u) = a exp(bu) then b = log(f(t)/f(s)) / (t-s) for each s < t, so
+            # \int_s^t f(u) du = (a/b) \int_s^t exp(bu) b du = (a/b)(exp(bt) - exp(bs))
+            #    = (t - s) * (f(t) - f(s)) / log(f(t) / f(s))
+            # unless b = 0, of course.
+            assert(steps[0] == 0)
```

Comment:
Nit picking here, but ``assert`` isn't a function. Should write ``assert steps[0] == 0``.

Reply:
I always forget which.